lib/core/src/Cardano/Wallet/DB/Sqlite/AddressBook.hs

```diff
@@ -162,41 +154,33 @@ emptyPool pool = pool{ Seq.indexedKeys = Map.empty }
     Shared key address book
 -------------------------------------------------------------------------------}
 -- | Isomorphism for multi-sig address book.
-instance
-    ( MkKeyFingerprint key (Proxy n, key 'AddressK XPub)
-    , MkKeyFingerprint key Address
-    , GetPurpose key
-    , SoftDerivation key
-    , Typeable n
-    , key ~ SharedKey
-    ) => AddressBookIso (Shared.SharedState n key)
+instance ( key ~ SharedKey ) => AddressBookIso (Shared.SharedState n key)
   where
     data Prologue (Shared.SharedState n key)
         = SharedPrologue (Shared.SharedState n key)
         -- Trick: We keep the type, but we empty the discovered addresses
     data Discoveries (Shared.SharedState n key)
-        = SharedDiscoveries (SeqAddressList 'UtxoExternal)
+        = SharedDiscoveries
+            ( Map
+                (KeyFingerprint "payment" SharedKey)
+                (Index 'Soft 'ScriptK, AddressState)
+            )
 
     addressIso = iso from to
       where
-        from st@(Shared.SharedState a b) = case b of
-            Shared.PendingFields{} ->
-                ( SharedPrologue st
-                , SharedDiscoveries (SeqAddressList [])
-                )
-            Shared.ReadyFields pool ->
-                let b0 = Shared.ReadyFields $ emptyPool pool
-                in  ( SharedPrologue (Shared.SharedState a b0)
-                    , SharedDiscoveries (toDiscoveriesShared @n pool)
+        from st = case Shared.ready st of
+            Shared.Pending -> (SharedPrologue st, SharedDiscoveries Map.empty)
+            Shared.Active pool ->
+                let pool0 = AddressPool.clear pool
+                in  ( SharedPrologue st{ Shared.ready = Shared.Active pool0 }
+                    , SharedDiscoveries $ AddressPool.addresses pool
                     )
-        to  ( SharedPrologue st@(Shared.SharedState a b0)
-            , SharedDiscoveries addrs
-            )
-          = case b0 of
-            Shared.PendingFields{} -> st
-            Shared.ReadyFields pool0
-                -> Shared.SharedState a $ Shared.ReadyFields
-                $ fromDiscoveries @n pool0 addrs
+        to (SharedPrologue st, SharedDiscoveries addrs)
+          = case Shared.ready st of
+            Shared.Pending -> st
+            Shared.Active pool0 ->
+                let pool = AddressPool.loadUnsafe pool0 addrs
```

Comment:
Why does the database need this backdoor? Is there a situation where we shouldn't fail if we read a not consistent address pool from the database?

Reply:
I would also prefer to use `AddressPool.load` here, but the idea is that the `Discoveries s` will be paired with `Prologue s` frequently, and outside the database layer. Then, it's important that the isomorphism `s ~ (Prologue s, Discoveries s)` is fast. I agree that `load` should be used near the database layer, but that is for another PR, as well as the issue that I would prefer to not export the constructors for `Prologue` and `Discoveries`.